File: codejam/codejam-pattern-matching.cpp
//
// Created by TianHongZXY on 2020/4/11.
//
#include <bits/stdc++.h>
using namespace std;

bool substr_(string l, string s){
    int len = s.length();
    int n = l.length();
    for(int i = 0; i <= n - len; i++){
        string sb = l.substr(i, len);
        if(sb.compare(s) == 0) return true;
    }
    return false;
}

int main(){
    int T, n, kase=0;
    string p;
    scanf("%d", &T);
    while(T--){
        scanf("%d", &n);
        getchar();
        bool possible = true;
        int len = 0;
        string candi;
        for(int i = 0; i < n; i++){
            getline(cin, p);
            p = p.substr(1);
            if(p.length() > len) {
                if (substr_(p, candi)) {
                    candi = p;
                    len = p.length();
                }
                else {
                    possible = false;
                    break;
                }
            }
            else {
                if (!substr_(candi, p)) {
                    possible = false;
                    break;
                }
            }
        }
        printf("Case #%d: ", ++kase);
        if(possible) cout << candi << endl;
        else cout << "*" << endl;
    }
    return 0;
}

File: codejam/codejam-pattern-matching.py
if __name__ == '__main__':
    T = int(input())
    kase = 1
    while(T):
        T -= 1
        length = 0
        candi = ""
        possible = True
        n = int(input())
        for i in range(n):
            line = input()
            line = line[1:]
            if(candi in line or line in candi):
                if(len(line) > length):
                    length = len(line)
                    candi = line
            else:
                possible = False
        print("Case #%d: " % (kase), end="")
        kase += 1
        if(possible): print(candi)
        else: print("*")

